test/gtest/ucm/cuda_hooks.cc

```diff
@@ -115,14 +78,90 @@ class cuda_hooks : public ucs::test {
             void *ptr, size_t size,
             ucs_memory_type_t expect_mem_type = UCS_MEMORY_TYPE_CUDA)
     {
-        ASSERT_EQ(ptr, alloc_event.address);
-        ASSERT_GE(alloc_event.size, size);
-        EXPECT_TRUE((alloc_event.mem_type == expect_mem_type) ||
-                    (alloc_event.mem_type == UCS_MEMORY_TYPE_UNKNOWN));
+        check_event_present(m_alloc_events, "alloc", ptr, size,
+                            expect_mem_type);
+    }
+
+    void check_mem_free_events(void *ptr, size_t size)
+    {
+        check_event_present(m_free_events, "free", ptr, size);
+    }
+
+private:
+    struct mem_event {
+        void              *address;
+        size_t            size;
+        ucs_memory_type_t mem_type;
+    };
+
+    void
+    check_event_present(const std::vector<mem_event> &events,
+                        const std::string &name, void *ptr, size_t size,
+                        ucs_memory_type_t mem_type = UCS_MEMORY_TYPE_UNKNOWN)
+    {
+        for (auto e : events) {
+            if (/* Start address match */
+                (ptr >= e.address) &&
+                /* End address match */
+                (UCS_PTR_BYTE_OFFSET(ptr, size) <=
+                 UCS_PTR_BYTE_OFFSET(e.address, e.size)) &&
+                /* Memory type match */
+                ((e.mem_type == mem_type) ||
+                 (e.mem_type == UCS_MEMORY_TYPE_UNKNOWN))) {
+                return;
+            }
+        }
+
+        FAIL() << "Could not file memory " << name << " event for " << ptr
+               << ".." << UCS_PTR_BYTE_OFFSET(ptr, size) << " type "
+               << ucs_memory_type_names[mem_type];
+    }
+
+    void mem_alloc_event(void *address, size_t size, ucs_memory_type_t mem_type)
+    {
+        m_alloc_events.push_back({address, size, mem_type});
+    }
+
+    void mem_free_event(void *address, size_t size)
+    {
+        m_free_events.push_back({address, size, UCS_MEMORY_TYPE_UNKNOWN});
+    }
+
+    static void cuda_vm_mapped_callback(ucm_event_type_t event_type,
+                                        ucm_event_t *event, void *arg)
+    {
+        cuda_hooks *self = reinterpret_cast<cuda_hooks*>(arg);
```

Comment:
```suggestion
        auto self = reinterpret_cast<cuda_hooks*>(arg);
```
Here and below. Because the type is already mentioned in cast parameter.